mypy/build.py

```diff
@@ -861,59 +863,134 @@ def is_file(path: str) -> bool:
+            if isdir:
+                dirs.append(dir)
+        find_module_dir_cache[dir_chain, lib_path] = dirs
+    third_party_dirs = []
+    # Third-party stub/typed packages
+    for pkg_dir in site_packages_dirs:
+        stub_name = components[0] + '-stubs'
+        typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
+        stub_dir = os.path.join(pkg_dir, stub_name)
+        if os.path.isdir(stub_dir):
+            stub_components = [stub_name] + components[1:]
+            path = os.path.join(pkg_dir, *stub_components[:-1])
+            if os.path.isdir(path):
+                third_party_dirs.append(path)
+        elif os.path.isfile(typed_file):
+            path = os.path.join(pkg_dir, dir_chain)
+            third_party_dirs.append(path)
+
+    return tuple(third_party_dirs +
+                 find_module_dir_cache[dir_chain, lib_path])
+
+
+def find_module_in_base_dirs(id: str, candidate_base_dirs: Iterable[str],
+                             last_component: str) -> Optional[str]:
+    # If we're looking for a module like 'foo.bar.baz', then candidate_base_dirs now
+    # contains just the subdirectories 'foo/bar' that actually exist under the
+    # elements of lib_path.  This is probably much shorter than lib_path itself.
+    # Now just look for 'baz.pyi', 'baz/__init__.py', etc., inside those directories.
+    seplast = os.sep + last_component
+    sepinit = os.sep + '__init__'
+    for base_dir in candidate_base_dirs:
+        base_path = base_dir + seplast  # so e.g. '/usr/lib/python3.4/foo/bar/baz'
+        # Prefer package over module, i.e. baz/__init__.py* over baz.py*.
+        for extension in PYTHON_EXTENSIONS:
+            path = base_path + sepinit + extension
+            path_stubs = base_path + '-stubs' + sepinit + extension
+            if is_file(path) and verify_module(id, path):
+                return path
+            elif is_file(path_stubs) and verify_module(id, path_stubs):
+                return path_stubs
+        # No package, look for module.
+        for extension in PYTHON_EXTENSIONS:
+            path = base_path + extension
+            if is_file(path) and verify_module(id, path):
+                return path
+    return None
+
+
+def find_module(id: str, lib_path_arg: Iterable[str],
+                python_executable: Optional[str]) -> Optional[str]:
     """Return the path of the module source file, or None if not found."""
     lib_path = tuple(lib_path_arg)
-
-    def find() -> Optional[str]:
-        # If we're looking for a module like 'foo.bar.baz', it's likely that most of the
-        # many elements of lib_path don't even have a subdirectory 'foo/bar'.  Discover
-        # that only once and cache it for when we look for modules like 'foo.bar.blah'
-        # that will require the same subdirectory.
+    if (id, python_executable, lib_path) not in find_module_cache:
         components = id.split('.')
         dir_chain = os.sep.join(components[:-1])  # e.g., 'foo/bar'
```

Comment:
nit: may as well do this line within `find_base_dirs`, since you don't use the result anywhere else